js/src/components/paid-ads/asset-group/images-selector.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
+import GridiconPlusSmall from 'gridicons/dist/plus-small';
+import GridiconCrossCircle from 'gridicons/dist/cross-circle';
+
+/**
+ * Internal dependencies
+ */
+import AppButton from '.~/components/app-button';
+import useCroppedImageSelector from '.~/hooks/useCroppedImageSelector';
+import './images-selector.scss';
+
+/**
+ * @typedef {Object} AssetImageConfig
+ * @property {number} minWidth The minimum width.
+ * @property {number} minHeight The minimum height.
+ * @property {number} suggestedWidth The suggested width.
+ * @property {number} suggestedHeight The suggested height.
+ */
+
+/**
+ * Renders a selector for asset images.
+ *
+ * @param {Object} props React props.
+ * @param {AssetImageConfig} props.imageConfig The config of the asset image.
+ * @param {string[]} props.initialImageUrls The initial image URLs.
+ * @param {number} [props.maxNumberOfImages=0] The maximum number of images. 0 by default and it means unlimited number.
+ */
+export default function ImagesSelector( {
+	imageConfig,
+	initialImageUrls = [],
+	maxNumberOfImages = 0,
+} ) {
+	const [ replacingImage, setReplacingImage ] = useState( null );
+	const [ images, setImages ] = useState( () =>
+		// The asset images fetched from Google Ads are only URLs.
+		initialImageUrls.map( ( url ) => ( { url, id: url, alt: '' } ) )
+	);
+
+	const removeImage = ( deletedImage ) => {
+		setImages( images.filter( ( { id } ) => id !== deletedImage.id ) );
+	};
+
+	const handle = useCroppedImageSelector( {
+		...imageConfig,
+		onDelete: removeImage,
+		onSelect( image ) {
+			const nextImages = [ ...images ];
+
+			// Find if there is a duplicate image first.
+			let index = nextImages.findIndex( ( { id } ) => id === image.id );
+
+			if ( replacingImage ) {
+				if ( index !== -1 && image.id !== replacingImage.id ) {
+					// If the selected image already exists while replacing, it's considered a swap position.
+					nextImages.splice( index, 1, { ...replacingImage } );
+				}
+				// If index gets -1 here, it means the image to be replaced has been deleted.
+				index = nextImages.indexOf( replacingImage );
```

Comment:
❓  I guess this means the we clicked on "Delete permanently" in the media selector?

Reply:
Yes, but ImagesSelector should be agnostic to the WP Media library, which relies only on the interface of useCroppedImageSelector. I adjusted the code comment a little bit in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/0932206b4476c5bce5080f0798f897af380243a9. Does this make it clearer?

Reply:
> Yes, but ImagesSelector should be agnostic to the WP Media library,

Could you elaborate this? Does it mean we can use another media Library rather the WP Media like a 3rd party one?

Reply:
❓  If so, should we test the PR against one 3rd party media library to see if it works?

Reply:
> > Yes, but ImagesSelector should be agnostic to the WP Media library,
> 
> Could you elaborate this? Does it mean we can use another media Library rather the WP Media like a 3rd party one?

It's simply due to the management/practice of interdependence.

Since the JS part of WP Media library is based on jQuery, Backbone.js and some `wp.*` libraries, and it also has its own lifecycle and state management, one of the purposes is to isolate the dependency on WP Media in `useCroppedImageSelector`.

`ImagesSelector`, as a hook user of `useCroppedImageSelector`, relies only on the interfaces provided by the hook for both code implementation and comments, and does not need to be coupled with the WP Media library. For example, `ImagesSelector` doesn't need to know how an image would be deleted, or by whom, but it only needs to know when and which image was deleted.

Therefore, that's the thought of "ImagesSelector should be agnostic to the WP Media library".

The possibility of replacing the implementation of `useCroppedImageSelector` with some other library is the result of this design approach, but not the purpose. In this case, I didn't expect WP Media would be replaced by a 3rd-party library, because in practice this probably won't happen.

Reply:
I adjusted the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/1819/commits/4b111e86bdd3c1869b50316094daae730d7e8273 since the -1 `index` case originally described no longer occurs after the fix in adb4734264ea9ac39e5c87c11b0229c6952757e0.